polkadot/runtime/parachains/src/assigner_on_demand/mod.rs

```diff
@@ -430,175 +685,140 @@ where
 	/// Adds an order to the on demand queue.
 	///
 	/// Paramenters:
-	/// - `order`: The `EnqueuedOrder` to add to the queue.
 	/// - `location`: Whether to push this entry to the back or the front of the queue. Pushing an
 	///   entry to the front of the queue is only used when the scheduler wants to push back an
 	///   entry it has already popped.
-	/// Returns:
-	/// - The unit type on success.
-	///
-	/// Errors:
-	/// - `InvalidParaId`
-	/// - `QueueFull`
 	fn add_on_demand_order(
-		order: EnqueuedOrder,
+		queue_status: &mut QueueStatusType,
+		para_id: ParaId,
 		location: QueuePushDirection,
-	) -> Result<(), DispatchError> {
-		// Only parathreads are valid paraids for on the go parachains.
-		ensure!(<paras::Pallet<T>>::is_parathread(order.para_id), Error::<T>::InvalidParaId);
-
-		let config = <configuration::Pallet<T>>::config();
-
-		OnDemandQueue::<T>::try_mutate(|queue| {
-			// Abort transaction if queue is too large
-			ensure!(
-				Self::queue_size() < config.scheduler_params.on_demand_queue_max_size,
-				Error::<T>::QueueFull
-			);
-			match location {
-				QueuePushDirection::Back => queue.push_back(order),
-				QueuePushDirection::Front => queue.push_front(order),
-			};
-			Ok(())
-		})
+	) {
+		let idx = match location {
+			QueuePushDirection::Back => queue_status.push_back(),
+			QueuePushDirection::Front => queue_status.push_front(),
+		};
+
+		let affinity = ParaIdAffinity::<T>::get(para_id);
+		let order = EnqueuedOrder::new(idx, para_id);
+		#[cfg(test)]
+		log::debug!(target: LOG_TARGET, "add_on_demand_order, order: {:?}, affinity: {:?}, direction: {:?}", order, affinity, location);
+
+		match affinity {
+			None => FreeEntries::<T>::mutate(|entries| entries.push(order)),
+			Some(affinity) =>
+				AffinityEntries::<T>::mutate(affinity.core_index, |entries| entries.push(order)),
+		}
 	}
 
-	/// Get the size of the on demand queue.
+	/// Decrease core affinity for para and update queue
 	///
-	/// Returns:
-	/// - The size of the on demand queue.
-	fn queue_size() -> u32 {
-		let config = <configuration::Pallet<T>>::config();
-		match OnDemandQueue::<T>::get().len().try_into() {
-			Ok(size) => return size,
-			Err(_) => {
-				log::debug!(
-					target: LOG_TARGET,
-					"Failed to fetch the on demand queue size, returning the max size."
-				);
-				return config.scheduler_params.on_demand_queue_max_size
-			},
+	/// if affinity dropped to 0, moving entries back to `FreeEntries`.
+	fn decrease_affinity_update_queue(para_id: ParaId, core_index: CoreIndex) {
+		let affinity = Pallet::<T>::decrease_affinity(para_id, core_index);
+		#[cfg(not(test))]
+		debug_assert_ne!(
+			affinity, None,
+			"Decreased affinity for a para that has not been served on a core?"
+		);
```

Comment:
first time seeing debug assert for non test environment. is the intention to run this assert on testnets and benchmarks only?

Reply:
I guess it breaks some edge case tests?